app/Models/Foundation/Summit/Registration/PromoCodes/SummitRegistrationPromoCode.php

```diff
@@ -667,4 +701,74 @@ public function clearTags()
     {
         $this->tags->clear();
     }
+
+    /**
+     * @return SummitAttendeeTicket[]
+     */
+    public function getTickets()
+    {
+        return $this->tickets;
+    }
+
+    /**
+     * @return SummitAttendeeTicket[]
+     */
+    public function getUnassignedTickets()
+    {
+        $criteria = Criteria::create();
+        $criteria->where(Criteria::expr()->isNull('owner'));
+        return $this->tickets->matching($criteria);
+    }
+
+    /**
+     * @param SummitAttendeeTicket $ticket
+     */
+    public function addTicket(SummitAttendeeTicket $ticket): void
+    {
+        if ($this->tickets->contains($ticket)) return;
+        $ticket->setPromoCode($this);
+        $this->tickets->add($ticket);
+    }
+
+    /**
+     * @param SummitTicketType $ticket_type
+     * @return SummitAttendeeTicket|null
+     */
+    public function getNextAvailableTicketPerType(SummitTicketType $ticket_type): ?SummitAttendeeTicket
+    {
+        $excluded_ids = [];
+
+        do {
+            $query = <<<SQL
```

Comment:
this is not SQL , its DQL , change <<<SQL to <<<DQL

Reply:
done